### DISCUSSION TITLE
Emacs rendering of combining characters

### DISCUSSION THREAD

**Discussion by sje30** [harfbuzz/harfbuzz#2790] [Q&A] [answered]
Hi,
Thank you for a great tool.  I'm using Emacs 27.1 from manjaro, which comes compiled with harfbuzz support.  I've found an issue with rendering of diacritials (notably x0305 for combining overline).

A description of the issue has been placed on the JuliaMono font issue page already,
https://github.com/cormullion/juliamono/issues/87

 where you can see I can't get proper rendering of "x bar" (the mean of x, in statistical terms) in the JuliaMono font in Emacs, but it does work in both xfce4-terminal and Chromium, which I believe also uses harfbuzz.  It also works for Monaco font in Emacs.  I therefore suspect Emacs.

I'm strugglling a bit to work out how best to debug this rendering, any hints appreciated! 

And please of course close this immediately if you think it noise!

Thanks very much

Stephen

**Comment by jfkthame** [comment, marked as answer]
My guess is that it's working in cases where the font has a precomposed glyph such that the (base letter, diacritic) pair gets mapped -- either by Unicode normalization within harfbuzz or by a GSUB ligature rule e.g. in the 'ccmp' feature -- to a single glyph; but fails when no precomposed glyph is present, so that the font is relying on mark positioning to place the diacritic correctly on the base letter.

In this case, there are still -- after shaping -- two glyphs to be rendered, and I suspect that somewhere in the emacs rendering code there's an insistence that each glyph needs to take up one character position in the display, so it effectively overrides the mark positioning that harfbuzz would have done.

Some fonts have a lot more precomposed accented forms than others, so which combinations fail would depend on the specific font being used.

(Note that the above is speculation; I have not tested or investigated it.)

**Reply by sje30** [reply]
thank you.  I've much to learn about this issue.

**Comment by khaledhosny** [comment]
You need to report this issue to Emacs, as there is no HarfBuzz issue here:
```
$ hb-view ~/Downloads/JuliaMono-Regular.ttf 'áâãāa̅ăȧäảåa̋ǎa̍a̎ȁa̐ȃa̒a̓a̔'
```
![](https://user-images.githubusercontent.com/93914/103349740-553cb300-4aa6-11eb-8355-079c7fb8518d.png)

```
$ hb-shape ~/Downloads/JuliaMono-Regular.ttf 'áâãāa̅ăȧäảåa̋ǎa̍a̎ȁa̐ȃa̒a̓a̔' --no-positions --no-clusters
[aacute|acircumflex|atilde|amacron|a|uni0305|abreve|uni0227|adieresis|uni1EA3|aring|a|uni030B|uni01CE|a|uni030D|a|uni030E|uni0201|a|uni0310|uni0203|a|uni0312|a|uni0313|a|uni0314]
```

Either it is the issue @jfkthame is guessing, or Emacs not using HarfBuzz as all for these character (Emacs does not use HarfBuzz unconditionally, it has some weird idea that only certain character sequences need to be passed to HarfBuzz).

**Reply by sje30** [reply]
thank you.

**Comment by jhcloos** [comment]
where it fails is with fonts like dejavu mono where the combining glyphs have the same width as the rest of the (mono) font, rather than zero width.

the result is a single glyph which is twice as wide and looks like two glyphs, buts acts as one for things like cursor movements.

w/in emacs this is a regression from how m17n works.

which they ignore because it is a hb issue.

hb needs to overlay combining glyphs even if they have +ve width, like previous libs do.

**Reply by khaledhosny** [reply]
HarfBuzz is working fine with DejaVuSansMono (except for U+0312 which is a font bug):
```
$ hb-view ~/Downloads/dejavu-fonts-ttf-2.37/ttf/DejaVuSansMono.ttf 'áâãāa̅ăȧäảåa̋ǎa̍a̎ȁa̐ȃa̒a̓a̔'
```
![](https://user-images.githubusercontent.com/93914/103352418-def07e80-4aae-11eb-9abc-ee27b9c70c51.png)

**Comment by jhcloos** [comment]
for comparison, this should render as a single "glyph", and does elsewhere and even weith hb using fonts like DejaVuSerif:

    :; hb-shape /usr/share/fonts/dejavu/DejaVuSansMono.ttf '0̸'
    [zero=0+1233|uni0338=0+1233]

vs:

    :; hb-shape /usr/share/fonts/dejavu/DejaVuSerif.ttf '0̸' 
    [zero=0+1303|uni0338=0+0]

mono's uni0338 ought also to say something like =0+0 there, even though the glyph has +ve width, simply because it is a combining CHARACTER.

**Reply by khaledhosny** [reply]
That is a font bug, the glyph for U+0338 has the wrong GDEF glyph class in DejaVuSansMono (base glyph instead of mark glyph) and no GPOS anchor attachment, same issue as U+0312 above.

**Reply by sje30** [reply]
I looked to file a bug report for the font, but can't see where to report it.  https://github.com/dejavu-fonts/dejavu-fonts seems to be inactive.

**Reply by khaledhosny** [reply]
That is the place, but indeed DejaVu fonts are abandoned by now.

**Reply by sje30** [reply]
would you happen to recommend other monospace fonts that have great Unicode support for maths?  (This is my original motivation for trying JuliaMono)

**Reply by khaledhosny** [reply]
No, I don’t. Sorry.

**Comment by jhcloos** [comment]
the a̒ in khaled’s note also shows this bug when one uses dejavumono for monospace….

i see it above as though it were a’.

**Comment by sje30** [comment]
I'm not quite following all the finer points here, but if everyone agrees here it is an Emacs bug, I will report it.  Thank you all for investigating.

**Reply by khaledhosny** [reply]
Your original issue is an Emacs bug.

**Reply by sje30** [reply]
great, thanks for confiming.

**Comment by sje30** [comment]
I have now submitted a bug report to Emacs.  ~~It may take up to 24 hours to appear on: https://lists.gnu.org/archive/html/bug-gnu-emacs/~~

https://debbugs.gnu.org/cgi/bugreport.cgi?bug=45557

**Reply by Eli-Zaretskii** [reply]
FTR, there doesn't seem to be any bug in Emacs in this regard, as the problem couldn't be reproduced on 2 different systems other than that of the OP.  My best current guess is that this is due to some local configuration or software version on the system where the problem exists.

**Reply by sje30** [reply]
I got this working, but through a route I don't understand!

Thank you everyone for your help.

https://lists.gnu.org/archive/html/bug-gnu-emacs/2021-01/msg00001.html

**Comment by jhcloos** [comment]
>>>>> "KH" == Khaled Hosny <notifications@github.com> writes:

KH> HarfBuzz is working fine with DejaVuSansMono (except for U+0312 which is a font bug):
KH> ```
KH> $ hb-view ~/Downloads/dejavu-fonts-ttf-2.37/ttf/DejaVuSansMono.ttf 'áâãāa̅ăȧäảåa̋ǎa̍a̎ȁa̐ȃa̒a̓a̔'
KH> ```

no it is not.

a̒ renders as one double-wide glyph which looks like an a followed by the
accent.

**Reply by behdad** [reply]
That's what Khaled meant when he said "except for U+0312 which is a font bug".

**Comment by jhcloos** [comment]
your png file confirms my local observation.

with fonts like dejavu serif, where the combining accent as width 0, hb overlays said accent on the base glyph.

not so with fonts like dejavu sans mono which have combining glyphs with positive width

**Reply by behdad** [reply]
> your png file confirms my local observation.
> 
> with fonts like dejavu serif, where the combining accent as width 0, hb overlays said accent on the base glyph.
> 
> not so with fonts like dejavu sans mono which have combining glyphs with positive width

Sigh. HarfBuzz has no problem with fonts like dejavu sans mono. The one sequence you have observed as not rendering correctly is a bug in this particular font.
